lib/web_catalog/api_extractor.es6.js

```diff
@@ -140,511 +242,428 @@ foam.CLASS({
-                while (!og.isType(prototypePropertyId) &&
-                       prototypePropertyId !== objectPrototypePropertyId &&
-                       prototypePropertyClassName === className) {
-                  prototypePropertyProps = prototypePropertyProps.concat(
-                      this.getObjectProperties_(og, prototypePropertyId));
-                  prototypePropertyId = og.getPrototype(prototypePropertyId);
-                  prototypePropertyClassName =
-                      this.getClassName_(prototypePropertyId, og);
+        }
+
+        // Gather APIs from libraries function/prototype pairs.
+        for (const id of allOgIds) {
+          const isCtor = this.fns_.hasOwnProperty(id);
+          const isLib = this.libs_.hasOwnProperty(id);
+          const isCtorProto = this.protos_.hasOwnProperty(id);
+
+          foam.assert(!((isCtor && isLib) || (isCtor && isCtorProto) ||
+                        (isLib && isCtorProto)),
+                      'ApiExtractor: Object with multiple personalities');
+
+          // Functions are interfaces, not prototypes. Just store properties and
+          // carry on.
+          if (isCtor) {
+            this.storeAPIsFromCtorOrLib_(og, id);
+          } else if (isLib) {
+            this.storeAPIsFromCtorOrLib_(og, id);
+            // Treat library as interface where Ctor = Ctor.prototype.
+            const ctorId = id;
+            const ctorNames = this.libs_[id];
+            let protoId = id;
+            let protoNames = this.getClassNames_(og, protoId);
+            this.storeProtoAPIsForCtor_(og, ctorId, ctorNames, protoId,
+                                        protoNames);
+          } else if (isCtorProto) {
+            // This is a "Foo.prototype", or a library object. Store its
+            // properties under "Foo", and pull up any of its prototype's
+            // properties, so long as those prototypes are not some
+            // "Bar.prototype".
+            const ctorIds = this.protos_[id];
+            for (const ctorId of ctorIds) {
+              const ctorNames = this.getFunctionNames_(og, ctorId);
+              let protoId = id;
+              let protoNames = this.getClassNames_(og, protoId);
+              this.storeProtoAPIsForCtor_(og, ctorId, ctorNames, protoId,
+                                          protoNames);
+            }
+          } // else: This is an instance or a primitive value.
+        }
+
+        // Copy APIs from instances down to closest interface prototype iff
+        // the API does not exist further down in the prototype chain.
+        for (const id of allOgIds) {
+          const isInterfacePrototype =
+              id => !og.isType(id) &&
+                  id !== this.objectDotPrototype_ &&
+                  id !== this.functionDotPrototype_;
+          const isCtor = this.fns_.hasOwnProperty(id);
+          const isCtorProto = this.protos_.hasOwnProperty(id);
```

Comment:
Does this not need to filter out libraries? The documentation says it is supposed to.

Reply:
Good point. It actually is supposed to include libraries because sometimes properties on the global object are "interface-defining instances". They are filtered out during the second pass at instances to avoid stripping away library APIs that were (also) copied down to an interface.